napari/layers/_layer_actions.py

```diff
@@ -0,0 +1,152 @@
+from functools import partial
+from typing import TYPE_CHECKING, Callable, Dict, Optional
+
+from typing_extensions import TypedDict
+
+from napari.experimental import link_layers, unlink_layers
+from napari.layers.utils._link_layers import get_linked_layers
+
+from .utils import stack_utils
+
+if TYPE_CHECKING:
+    from napari.components import LayerList
+
+
+def _duplicate_layer(ll: LayerList):
+    for lay in list(ll.selection):
+        new = deepcopy(lay)
+        new.name += ' copy'
+        ll.insert(ll.index(lay) + 1, new)
+
+
+def _split_stack(ll: LayerList, axis: int = 0):
+    layer = ll.selection.active
+    if not layer:
+        return
+    if layer.rgb:
+        images = stack_utils.split_rgb(layer)
+    else:
+        images = stack_utils.stack_to_images(layer, axis)
+    ll.remove(layer)
+    ll.extend(images)
+    ll.selection = set(images)  # type: ignore
+
+
+def _convert(ll: LayerList, type_: str):
+    from .base.base import Layer
+
+    for lay in list(ll.selection):
+        idx = ll.index(lay)
+        data = lay.data.astype(int) if type_ == 'labels' else lay.data
+        ll.pop(idx)
+        ll.insert(idx, Layer.create(data, {'name': lay.name}, type_))
+
+
+def _merge_stack(layer_list, rgb=False):
+    selection = list(layer_list.selection)
+    for layer in selection:
+        layer_list.remove(layer)
+    if rgb:
+        new = stack_utils.merge_rgb(selection)
+    else:
+        new = stack_utils.images_to_stack(selection)
+    layer_list.append(new)
+
+
+class ContextAction(TypedDict):
+    """A dict that encapsulates a QAction in a QtActionContextMenu.
+
+    Parameters
+    ----------
+    description : str
+        The words that appear in the menu
+    action : callable
+        A function that may be called if the item is selected in the menu
+    enable_when : str
+        An expression that evaluates to a boolean (in namespace of some
+        context) and controls whether the menu item is enabled.
+    show_when : str
+        An expression that evaluates to a boolean (in namespace of some
+        context) and controls whether the menu item is visible.
+    """
+
+    description: str
+    action: Callable
+    enable_when: str
+    show_when: Optional[str]
+
+
+# Each item in LAYER_ACTIONS will be added to the `QtActionContextMenu` created
+# in _qt.containers._layer_delegate.LayerDelegate (i.e. they are options in the
+# menu when you right-click on a layer in the layerlist.)
+#
+# variable names used in the `enable_when` and `show_when` expressions must be
+# keys in the napari.components.layerlist.CONTEXT_KEYS dict.  If you need a new
+# context paramameter, add a key key:value pair to the CONTEXT_KEYS dict.
+#
+# `action` must be a callable that accepts a single argument, an instance of
+# `LayerList`.
+
+LAYER_ACTIONS: Dict[str, ContextAction] = {
+    'napari:duplicate_layer': {
+        'description': 'Duplicate Layer',
+        'action': _duplicate_layer,
+        'enable_when': 'True',
```

Comment:
Can you add `'show_when': 'True'` to all the places where it is missing. I find the extra line removes all ambiguity - it's much easier for me to read that and know it's always shown then have to remember if nothing specified then it's always shown. I think I might require all 4 params, `'description', 'action', 'enable_when', 'show_when'` and throw an error if one is missing. Just keeps things simple

Reply:
I can, but it will incur an extra (unnecessary) `eval` for each one of them.  Probably negligible overhead, but that's why I left it out.  I'll test it

Reply:
made them all mandatory.  (No runtime error if one is missing, but added a test)